### DISCUSSION TITLE
Accessing Members of `RustOpaque` Objects from Dart

### DISCUSSION THREAD

**Discussion by coder0xff** [fzyzcjy/flutter_rust_bridge#986] [Ideas]
Having learned about [RustOpaque](https://cjycode.com/flutter_rust_bridge/feature/lang_rust_opaque.html), I wanted to sandbox a bit and see what it was all about. I wrote `api_sandbox.rs`.

```rust
pub fn f() -> RustOpaque<fn(i32) -> i32> {
    fn g(x: i32) -> i32 { return x*2; }
    return RustOpaque::<fn(i32) -> i32>::new(g);
}
```
I was hopeful that `RustOpaque<fn(i32) -> i32>` would result in a callable Dart type.

```dart
@sealed
class FnI32I32 extends FrbOpaque {
  final ApiSandbox bridge;
  FnI32I32.fromRaw(int ptr, int size, this.bridge) : super.unsafe(ptr, size);
  @override
  DropFnType get dropFn => bridge.dropOpaqueFnI32I32;

  @override
  ShareFnType get shareFn => bridge.shareOpaqueFnI32I32;

  @override
  OpaqueTypeFinalizer get staticFinalizer => bridge.FnI32I32Finalizer;
}
```

It doesn't seem to be callable, but hey it's called "Opaque" for a reason. I can write this, however.

```rust
pub fn invoke_i32_i32(func: RustOpaque<fn(i32) -> i32>, v: i32) -> i32 {
    (*func)(v)
}
```
Then in Dart,

```dart
var func = await apiSandbox.f();
var result = await apiSandbox.invokeI32I32(func: func, v: 6);
assert(result == 12);
```

I think flutter_rust_bridge could generate these helpers and make invocation transparent to callers. But why stop there? Since Dart supports getter and setter functions, *all* the fields and methods could be exposed on a "by reference" object, though probably best kept as a separate feature from `RustOpaque`. If very tightly integrated, it could do this to all non-serializable values by default, without any explicit encapsulation:

```rust
pub fn f() -> fn(i32) -> i32 {
    fn g(x: i32) -> i32 { return x*2; }
    return g;
}
```

```dart
var func = await apiSandbox.f();
var result = await func(6);
assert(result == 12);
```

I'm thinking of attempting it myself (just fn to start), but I wanted to reach out and ask if this has already been tried and if there's any standing knowledge on the feasibility. Thanks!

**Comment by fzyzcjy** [comment]
That looks interesting! 

I guess the more general form of this proposal is like, "exposing methods defined on opaque types". 

We already understand methods defined on non-opaque types https://cjycode.com/flutter_rust_bridge/feature/lang_methods.html. So does this work on opaque types? (or with minor modifications?)

cc @rogurotus who implemented the great opaque types
